### ISSUE TITLE
jfx_bridge List Slicing TypeError

### ISSUE THREAD

**Issue by 6r0k3d** [justfoxing/jfx_bridge#14]
Follow up from the ghidra_bridge comment:

jfx-bridge is now throwing a TypeError:
```
WARNING:jfx_bridge.bridge:Handling connection from ('127.0.0.1', 50000)
Traceback (most recent call last):
  File "C:\Users\student\Downloads\ghidra_10.0.4_PUBLIC_20210928\ghidra_10.0.4_PUBLIC\Ghidra\Features\Python\ghidra_scripts\jfx_bridge\bridge.py", line 974, in local_get
    print(name)
AttributeError: 'javapackage' object has no attribute '__all__'
Traceback (most recent call last):
  File "C:\Users\student\Downloads\ghidra_10.0.4_PUBLIC_20210928\ghidra_10.0.4_PUBLIC\Ghidra\Features\Python\ghidra_scripts\jfx_bridge\bridge.py", line 974, in local_get
    print(name)
AttributeError: 'javapackage' object has no attribute '__all__'
Traceback (most recent call last):
  File "C:\Users\student\Downloads\ghidra_10.0.4_PUBLIC_20210928\ghidra_10.0.4_PUBLIC\Ghidra\Features\Python\ghidra_scripts\jfx_bridge\bridge.py", line 974, in local_get
    print(name)
AttributeError: 'javapackage' object has no attribute '__all__'
Traceback (most recent call last):
  File "C:\Users\student\Downloads\ghidra_10.0.4_PUBLIC_20210928\ghidra_10.0.4_PUBLIC\Ghidra\Features\Python\ghidra_scripts\jfx_bridge\bridge.py", line 974, in local_get
    print(name)
AttributeError: 'javapackage' object has no attribute '__all__'
Traceback (most recent call last):
  File "C:\Users\student\Downloads\ghidra_10.0.4_PUBLIC_20210928\ghidra_10.0.4_PUBLIC\Ghidra\Features\Python\ghidra_scripts\jfx_bridge\bridge.py", line 1083, in local_call
    )
TypeError: array.array indices must be integers
WARNING:jfx_bridge.bridge:Closing connection from ('127.0.0.1', 50000)
```

I threw in some prints, but I'm not finding them useful, hopefully it makes more sense to you

```    
    @stats_hit
    def local_get(self, args_dict):
        handle = args_dict[HANDLE]
        name = args_dict[NAME]
        self.logger.debug("local_get: {}.{}".format(handle, name))

        target = self.get_object_by_handle(handle)
        print("target: ", target)
        print("name: ", name)

@stats_hit
    def local_call(self, args_dict):
        print("HANDLE: ", HANDLE)
        print(args_dict)
```

Output seems to print from local_get, but not local_call:
```
target:  <class 'slice'>
name:  __str__
target:  <class 'slice'>
name:  __hash__
target:  <class 'slice'>
name:  __eq__
target:  <class 'slice'>
name:  __ge__
target:  <class 'slice'>
name:  __gt__
target:  <class 'slice'>
name:  __le__
target:  <class 'slice'>
name:  __lt__
target:  <class 'slice'>
name:  __ne__
target:  <class 'slice'>
name:  __subclasshook__
```

**Comment by justfoxing** [comment]
Hmm, not enough info for me to see what's going on.

Can you narrow this down in the script you're running to work out what
expression is causing it, and let me know what that is? (The full script
would be handy as well)

Another handy debugging trick is to try running the same operations in
ghidra's internal python window - if the operation works there, it's a
bridge problem, but if you get the same error, it's a problem with your
script or ghidra itself.

**Comment by 6r0k3d** [comment]
Ok, just got this running in an IDE instead of notepad++ and have actual useful info-

my code runs the following:
`print("length inputs: ", len(inputs[1:]))`

looks like bridge doesnt like list slicing?

Thought this picture captured the info better then trying to individually copy / paste the Watch output:
![image](https://user-images.githubusercontent.com/42014669/153931033-729023ae-7f4f-4161-8044-6d4abc0ab205.png)

**Comment by 6r0k3d** [comment]
Ha, the timing on our posts.

Full script is here, I'm working on modifying GRIMM's GEARSHIFT to auto extract C++ classes: [GEARSHIFT](https://github.com/grimm-co/GEARSHIFT/blob/a532fc76e2be59092174e3a1aea2ab8f28ef4bf0/plugin/PCodeInterpreter.py#L423)

**Comment by justfoxing** [comment]
Cool, thanks for that. I've confirmed I can replicate the error, and that it doesn't happen in purely internal Ghidra python - think you're right and it's an issue with how slices work. I'm going to have to find some time to take a poke around and work out what's going on.

As a quick and dirty hack-around, you can fix this by casting the remote array to a local list (e.g. `inputs[1:]` would become `list(inputs)[1:]`) - the bridge will extract the content of the array into a python list, which will respect slicing. Obviously not the ideal solution (especially for large arrays :/), but it might stop you getting blocked for now, until I can work out an actual fix.

**Comment by justfoxing** [comment]
Fixed in jfx-bridge 0.9.0. 

What was happening was that slices created when the [1:] syntax was used were being pushed across the bridge as bridged objects. When they hit __getitem__ on the array on the other side, it expects either an int or something where the type is "slice" - but these were type "_bridged_slice". Now slices are serialized across the bridge as their start/stop/step values, and recreated as local slice objects on the other side. The type is what's expected, and __getitem__ is happy again. 

TL;DR - update and reinstall with `pip install jfx-bridge --upgrade --force-reinstall; python -m ghidra_bridge.install_server <script location>`, and you should be able to go back to doing `inputs[1:]` without the list() wrapping.
